### PR TITLE
Make `Out-String` and `Out-File` keep string input unchanged

### LINKED ISSUE #17452: Save pure string text to file with redirection or `Out-File` unexpectedly strip off VT sequences from the original text
### Prerequisites

- [X] Write a descriptive title.
- [X] Make sure you are able to repro it on the [latest released version](https://github.com/PowerShell/PowerShell/releases)
- [X] Search the existing issues.
- [X] Refer to the [FAQ](https://github.com/PowerShell/PowerShell/blob/master/docs/FAQ.md).
- [X] Refer to [Differences between Windows PowerShell 5.1 and PowerShell](https://docs.microsoft.com/powershell/scripting/whats-new/differences-from-windows-powershell).

### Steps to reproduce

This happens in 7.3.0-preview.4, and all other 7.3 preview versions.
**This is a regression from both 7.0.x and 7.2.x (both are LTS versions).**

Content of the test file `try.ps1`: (Ansi escape sequence is used for color).
```
Read-Host -Prompt "[33mEnter your device code[0m"
```

Try replacing ``"`r`n"`` with ``"`n"`` for the content in the file and save it to another file:
```pwsh
(Get-Content .\try.ps1 -Raw).Replace("`r`n", "`n") > new.ps1
## Or, (Get-Content .\try.ps1 -Raw).Replace("`r`n", "`n")  | Out-File new.ps1
```

Now all Ansi escape sequences are lost in the content of `new.ps1`:
```
Read-Host -Prompt "Enter your device code"
```
---------------------------------------------------------------

`Set-Content` works as expected, which preserves the Ansi escape sequences.

```pwsh
(Get-Content .\try.ps1 -Raw).Replace("`r`n", "`n") | Set-Content new.ps1
```
---------------------------------------------------------------

Behavior in 7.0.11:

image

Behavior in 7.2.2:

image

-------------------------------------------

### Root cause and behavior proposal

The root cause of this regression is having `Host` to be the default value for `$PSStyle.OutputRendering`, which make the `Out-*` cmdlet to strip off escape sequences from the passed-in input.

- In 7.0.x, the `$PSStyle` feature is not there, so `Out-*` commands don't do anything special to Ansi escape sequences in the input.
- In 7.2.x, the default value for `$PSStyle.OutputRendering` is `Ansi`, so the behavior is unchanged without a user knowing about the `$PSStyle` feature.
- In 7.3, the default value is changed to `Host`, which makes `Out-*` cmdlets to always strip off escape sequences from the input, and hence the regression.

I understand the purpose of having `Host` option for `$PSStyle.OutputRendering` -- it's nice to have ANSI decorations in the formatting output when it's rendered on terminal, but when redirecting the formatting out like in `Get-Process > file`, the escape sequences in the formatting output becomes annoying. The `Host` option is to tell PowerShell to not keep ANSI decorations in case of redirection (essentially `Out-File`, and basically any `Out-*` cmdlets such as `Out-String`).

But for pure string input that doesn't really go through formatting, it doesn't seem right to remove the Ansi escape sequences, given that the formatting system doesn't add any to it. The `Out-*` should treat the pure string input as is, and only remove escape sequences when the input object actually goes through formatting.

### Expected behavior

```console
The content in new.ps1 should be the same as in try.ps1, except that all "\r\n" characters are replaced with "\n".
```

### Actual behavior

```console
The escape sequences in the original file was removed during redirection.
```

### Error details

```console
N/A
```

### Environment data

```powershell
PS:24> $PSVersionTable

Name                           Value
----                           -----
PSVersion                      7.3.0-preview.4
PSEdition                      Core
GitCommitId                    7.3.0-preview.4
OS                             Microsoft Windows 10.0.19044
Platform                       Win32NT
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0…}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0
```

### Visuals

_No response_

### PR DESCRIPTION
**Pull request by daxian-dbw**
# PR Summary

Fix #17452

The changes in this PR update the behavior of `Out-File` and `Out-String` regarding the `RenderingOutput` setting to be the following:
- when the input object is pure string, these 2 cmdlets keep the string input unchanged regardless of the `RenderingOutput` setting
- when the input object needs to have formatting views applied to it, these 2 cmdlets remove escape sequences from the formatting output strings based on the `RenderingOutput` setting.

This change is a breaking change to these 2 cmdlets' existing behavior regarding `RenderingOutput`.

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [ ] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [ ] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/main/reference/7.3/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): 
- **User-facing changes**
    - [ ] Not Applicable
    - **OR**
    - [x] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [x] Issue filed: https://github.com/MicrosoftDocs/PowerShell-Docs/issues/8892
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [x] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.

### CODE REVIEW

**Review by iSazonov — APPROVED** [review]

**Comment by iSazonov** [comment]
We need to update docs https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_ansi_terminals?view=powershell-7.2

**Review by SeeminglyScience — APPROVED** [review]
LGTM! Thank you for making it still possible to retain formatting when output rendering is `Ansi` ❤️

**Review by TravisEz13 — APPROVED** [review]

**Review by SteveL-MSFT — CHANGES_REQUESTED** [review]

**Comment by ImportTaste** [comment]
Is this going to get rid of the extra linebreak that happens too?

File: test/powershell/engine/Formatting/PSStyle.Tests.ps1 [hunk 1]
```diff
@@ -245,6 +245,80 @@ Describe 'Tests for $PSStyle automatic variable' {
         $result = & $pwsh -noprofile -Command { $PSStyle.Progress.UseOSCIndicator = $true; 'hello'} | Format-List
         $result | Out-String -Stream | Should -BeExactly 'hello'
     }
+
+    It "String intput to Out-String should be intact" {
+        $oldRender = $PSStyle.OutputRendering
+        $testStr = "`e[31mABC`e[0m"
+
+        try {
+            $PSStyle.OutputRendering = 'Ansi'
```
[the remark this answers is not in the document]
Comment (daxian-dbw) [reply test/powershell/engine/Formatting/PSStyle.Tests.ps1]: Fixed

**Comment by daxian-dbw** [comment]
@ImportTaste This PR only affects escape sequences.

**Review by SteveL-MSFT — APPROVED** [review]
Really great change!

File: test/powershell/engine/Formatting/PSStyle.Tests.ps1 [hunk 2]
```diff
@@ -286,6 +286,77 @@ Describe 'Tests for $PSStyle automatic variable' {
...
+        Set-Content -Path $TestDrive\test.ps1 -Value $content -Encoding utf8NoBOM
+
+        try {
+            $PSStyle.OutputRendering = $OutputRendering
+            Get-Content $TestDrive\test.ps1 > $TestDrive\copy.ps1
+            (Get-Content $TestDrive\copy.ps1 -Raw).Trim() | Should -BeExactly $content
+        }
+        finally {
+            $PSStyle.OutputRendering = $oldRender
+            Remove-Item $TestDrive\test.ps1 -Force
```
Comment (JamesWTruher) [inline test/powershell/engine/Formatting/PSStyle.Tests.ps1:331]: these `Remove-Item` should not be needed. Files in `$TestDrive` are cleaned up by Pester automatically.

**Review by JamesWTruher — COMMENTED** [review]
I think we should seriously consider adding a parameter to out-string and out-file which retains our current behavior. Perhaps a parameter `-AsPlainText` or some such will allow users to have the older behavior while we provide a safer default behavior.

File: test/powershell/engine/Formatting/PSStyle.Tests.ps1 (hunk 2 above)
Comment (daxian-dbw) [reply test/powershell/engine/Formatting/PSStyle.Tests.ps1:331]: IIRC, Pester remove the whole `Test` drive after a `describe` finishes, but not individual files after ever `it` finishes.
Here I'm reusing the same file for 3 different `it` runs, so removing the file just to be safe.

**Comment by daxian-dbw** [comment]
I agree that we should consider so, but maybe wait until we get issue reports that someone really gets affected by this breaking change.
Also, `-AsPlainText` maybe misleading because it means "leaves the string as is" in other places where this parameter name is used, not "removing ANSI sequences", but we can discuss this more when we really need to consider adding this parameter.

**Comment by SteveL-MSFT** [comment]
@PowerShell/powershell-committee reviewed this, we would recommend adding a parameter to enable users to get back the previous behavior.

**Comment by daxian-dbw** [comment]
@SteveL-MSFT Is that a requirement for this PR or could do in separate PR? Also, is that supposed to be done now or should wait until there is a report about needing the previous behavior?

**Comment by daxian-dbw** [comment]
Given that the recommendation is a separate feature, I think it would be fine to use a separate PR if we really decided to add a parameter for the previous behavior. I will merge this PR and open an issue for that.

**Comment by adityapatwardhan** [comment]
Backport done via https://github.com/PowerShell/PowerShell/pull/17859

**Pull request merged**